util/btrie.js

```diff
@@ -0,0 +1,445 @@
+// -*- mode: js; indent-tabs-mode: nil; js-basic-offset: 4 -*-
+//
+// This file is part of Almond
+//
+// Copyright 2019 The Board of Trustees of the Leland Stanford Junior University
+//
+// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
+//
+// See COPYING for details
+"use strict";
+
+const assert = require('assert');
+
+const GrowableBuffer = require('./growable_buffer');
+
+const WILDCARD = {};
+
+const FILE_HEADER_LENGTH = 6;
+
+const NodeType = {
+    DATA: 1,
+    LEAF: 2,
+    COMPACT: 3,
+    INTERMEDIATE: 4,
+    NODE_TYPE_MAX: 4
+};
+const NodeFlags = {
+    NONE: 0,
+    WILDCARD: 8,
+};
+
+function writeNodeHeader(buffer, offset, nodeType, flags = NodeFlags.NONE) {
+    assert(nodeType >= 0 && nodeType <= NodeType.NODE_TYPE_MAX);
+    assert((flags & 0b111) === 0);
+    buffer.writeUInt8(nodeType | flags, offset);
+    offset++;
+    return offset;
+}
+
+function readNodeHeader(buffer, offset) {
+    const header = buffer.readUInt8(offset);
+    assert((header & 0b111) <= NodeType.NODE_TYPE_MAX);
+    const nodeType = header & 0b111;
+    const flags = header & ~0b111;
```

Comment:
what does `& 0b111` do?

Reply:
0b111 is 111 written in binary, so it takes the low 3 bits. The header is a 8 byte value, the low 3 bits is the node type (unsigned int) and the high bits are flags.